src/Raven.Server/Utils/CombinedReadingStream.cs

```diff
@@ -0,0 +1,138 @@
+﻿using System;
+using System.IO;
+
+namespace Raven.Server.Utils
+{
+    public class CombinedReadingStream : Stream
```

Comment:
This seems like it is very complex, mostly because it is a stream of streams. 
It is better if it was just managing the raw pointers directly, since that way, it would have to do a lot less work.